pkg/export/otel/traces.go

```diff
@@ -246,51 +252,74 @@ func (r *TracesReporter) reportClientSpan(span *transform.HTTPRequestSpan, trace
 		}
 	}
 
-	makeSpan(ctx, tracer, span)
+	r.makeSpan(ctx, tracer, span)
 }
 
 func (r *TracesReporter) reportServerSpan(span *transform.HTTPRequestSpan, tracer trace2.Tracer) {
-	s := makeSpan(r.ctx, tracer, span)
-	topSpans.Add(span.ID, s)
-	cs, ok := clientSpans.Get(span.ID)
-	newer := []transform.HTTPRequestSpan{}
-	if ok {
-		// finish any client spans that were waiting for this parent span
-		for j := range cs {
-			cspan := &cs[j]
-			if cspan.Inside(span) {
-				makeSpan(s.RootCtx, tracer, cspan)
-			} else if cspan.Start > span.RequestStart {
-				newer = append(newer, *cspan)
+
+	s := r.makeSpan(r.ctx, tracer, span)
+	if span.ID != 0 {
+		topSpans.Add(span.ID, s)
+		cs, ok := clientSpans.Get(span.ID)
+		newer := []transform.HTTPRequestSpan{}
+		if ok {
+			// finish any client spans that were waiting for this parent span
+			for j := range cs {
+				cspan := &cs[j]
+				if cspan.Inside(span) {
+					r.makeSpan(s.RootCtx, tracer, cspan)
+				} else if cspan.Start > span.RequestStart {
+					newer = append(newer, *cspan)
+				} else {
+					r.makeSpan(r.ctx, tracer, cspan)
+				}
+			}
+			if len(newer) == 0 {
+				clientSpans.Remove(span.ID)
 			} else {
-				makeSpan(r.ctx, tracer, cspan)
+				clientSpans.Add(span.ID, newer)
 			}
 		}
-		if len(newer) == 0 {
-			clientSpans.Remove(span.ID)
-		} else {
-			clientSpans.Add(span.ID, newer)
-		}
 	}
 }
 
 func (r *TracesReporter) reportTraces(input <-chan []transform.HTTPRequestSpan) {
 	defer r.close()
-	tracer := r.traceProvider.Tracer(reporterName)
+	defaultTracer := r.traceProvider.Tracer(reporterName)
 	for spans := range input {
 		for i := range spans {
+			spanTracer := defaultTracer
 			span := &spans[i]
 
+			if span.ServiceName != "" {
+				spanTracer = r.namedTracer(span.ServiceName)
+			}
+
 			switch span.Type {
 			case transform.EventTypeHTTPClient, transform.EventTypeGRPCClient:
-				r.reportClientSpan(span, tracer)
+				r.reportClientSpan(span, spanTracer)
 			case transform.EventTypeHTTP, transform.EventTypeGRPC:
-				r.reportServerSpan(span, tracer)
+				r.reportServerSpan(span, spanTracer)
 			}
 		}
 	}
 }
 
+func (r *TracesReporter) namedTracer(comm string) trace2.Tracer {
```

Comment:
I had to make something like this, since the Otel library won't let us name the service per span. Even though there's an option to set the service name in the attributes section, tempo wants it from the actual resource on the trace provider. If I didn't do this, we end up with 'unknown parent span' in Tempo events list.

Technically, we need to close all these cached providers, but the cached providers all share the batched span processor, which is the only thing that the provider closes. Therefore I think we are OK with only 1 close from the default trace provider.